aten/src/ATen/native/Linear.cpp

```diff
@@ -22,11 +172,13 @@ Tensor bilinear(const Tensor& input1, const Tensor& input2, const Tensor& weight
             "bilinear(): bias size does not match weight size: got %lld but expected %lld",
             (long long)bias.size(0), (long long)weight.size(0));
 
-  auto b_input1 = input1.unsqueeze(-2).unsqueeze(-2);
-  auto b_input2 = input2.unsqueeze(-2).unsqueeze(-1);
-
-  auto output = at::matmul(at::matmul(b_input1, weight), b_input2);
-  output = output.squeeze(-1).squeeze(-1);
+  std::vector<int64_t> output_size;
+  auto size1 = input1.sizes();
+  output_size.insert(output_size.end(), size1.begin(), size1.end() - 1);
+  output_size.push_back(weight.size(0));
+  auto input1_flattened = input1.view({-1, input1.size(-1)});
+  auto input2_flattened = input2.view({-1, input2.size(-1)});
+  Tensor output = at::_trilinear(input1_flattened, weight, input2_flattened, {1,3}, {0}, {1,2}, {2,3}).reshape(output_size);
```

Comment:
Can you explain what the difference between `bilinear` and `_trilinear` is? The code is quite difficult to comprehend, because it uses ops we didn't have previously, and they are undocumented, so it's hard to figure out what they do

Reply:
`bilinear` is `torch.nn.bilinear`, i.e. `bilinear_ba = sum_n sum_m left_bn weight_anm right_bm`

`_trilinear` is a generalisation in that you can specify arbitrary unsqueezing (or equivalently arbitrary indices in the factors above) and summation dimensions before multiplying the three factors. This is used to have a function that can express its own derivative in terms of itself in order to not have to define higher order derivatives explicitly. I would not know how to unify `bilinear` and `bilinear_backward` without `_trilinear`. The naming discrepancy bilinear vs. trilinear comes from viewing the weights as parameters (then it is bilinear in left and right) vs. inputs.

sumproduct_pair computes an einsum-style reduction after dimensions have been aligned.

I'll add comments in the Linear.cpp as well.